velox/expression/ExprCompiler.cpp

```diff
@@ -442,8 +442,13 @@ ExprPtr compileExpression(
     if (constant->hasValueVector()) {
       result = std::make_shared<ConstantExpr>(constant->valueVector());
     } else {
-      result = std::make_shared<ConstantExpr>(
-          BaseVector::createConstant(constant->value(), 1, pool));
+      if (constant->value().isNull()) {
```

Comment:
This will return a null constant for Opaque Types as opposed to failing.
See https://github.com/facebookincubator/velox/blob/main/velox/vector/BaseVector.cpp#L593

Reply:
It should be possible to specialize `newConstant` for complex type nulls?
https://github.com/facebookincubator/velox/blob/49ec747ccb6c4562949545c5bac463edfa3659ea/velox/vector/BaseVector.cpp#L611-L617

Reply:
Hi Deepak, Thank you for catching this! Adding a path for complex types in `createConstant` is actually not that easy, because `variant value` carries only the type kind, but not the exact type that can be nested (e.g., ARRAY(ARRAY(...))). On the other hand, there is indeed something wrong with the handling of Opaque type here as you pointed out. I discussed with @mbasmanova and we plan to fix it in `BaseVector::createNullConstant` separately. Does this sound good?

Reply:
@kagamiori sounds good. Thanks for clarifying.